CAMEL-15496: Headers and Properties support

rgannu opened the pull request:
The additional headers will be added on top of the existing headers
from the message.

The properties added are only basic AMQP properties as defined in the
{@link com.rabbitmq.client.AMQP.BasicProperties.Builder}. When the
message contains already these properties then will be considered and
these additional properties will be ignored.

Introduced a SimpleDataHolderBean to hold a hashmap for adding any
header. Changed the RabbitMQConstants to Enum so that we can also add
the property description in that.

As with this fix though we can add any header, only static/hard-coded
values could be associated.

The rabbitmq AMQP basic properties constant keys should not
have dot or hyphen in their prefixes. This gives problem in trying to
retrieve the value due to fix of CAMEL-15394. Changed to underscore.

- [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] If you're unsure, you can format the pull request title like `[CAMEL-XXX] Fixes bug in camel-file component`, where you replace `CAMEL-XXX` with the appropriate JIRA issue.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Run `mvn clean install -Psourcecheck` in your module with source check enabled to make sure basic checks pass and there are no checkstyle violations. A more thorough check will be performed on your pull request automatically.
Below are the contribution guidelines:
https://github.com/apache/camel/blob/master/CONTRIBUTING.md

rgannu commented:
@oscerd gave many comments in the other PR (which i had to close as there were repeated build failures.
https://github.com/apache/camel/pull/4241

Due to the fix of CAMEL-15394, the rabbitmq AMQP basic properties constant keys when it contains dots / hyphens doesn't work. Hence, I changed to underscore. In case if @davsclaus  can add support for dots back in the RabbitMQConstants, then I can remove the underscore from this PR.

omarsmak commented on catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/docs/rabbitmq-component.adoc:
```diff
@@ -292,37 +296,37 @@ camel exchange then they will be set on the RabbitMQ message.
+|`rabbitmq_REPLY_TO` |The replyTo to set on the RabbitMQ message
 
-|`rabbitmq.CONTENT_ENCODING` |The contentEncoding to set on the RabbitMQ message
+|`rabbitmq_CONTENT_ENCODING` |The contentEncoding to set on the RabbitMQ message
```
I am wondering, would it be better if you are changing these headers, to follow the Camel style naming, e.g: `CamelRabbitMqContentEncoding`? 
Also such a change, needs to be added in the migration guide as it is a breaking change for existing camel routes

rgannu replied:
I didn't want to even had to touch this Constants class but had to change as having dots in the constant value isn't working (I think this is after the introduction of CAMEL-15394). 
Also such a change, needs to be added in the migration guide as it is a breaking change for existing camel routes
>> This is what @oscerd was also mentioning. Then, I wud prefer that existing constant value also works now. 
Can we please fix that issue? Then, I do not have to adapt from dot to underscore. Thanks

omarsmak replied:
> I didn't want to even had to touch this Constants class but had to change as having dots in the constant value isn't working (I think this is after the introduction of [CAMEL-15394](https://issues.apache.org/jira/browse/CAMEL-15394)).

I see, then it was already not following Camel headers format (with the dot), hence I'd recommend to change these to follow Camel's format e.g `CamelRabbitmqContentEncoding` and put a note into the migration guide for the users since we are at it. @oscerd @davsclaus WDYT?

oscerd replied:
I'm ok with this. I think we can break the behavior until the next LTS

omarsmak commented on components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:
```diff
@@ -16,42 +16,87 @@
-
-    private RabbitMQConstants() {
-        // Constants class
+    ROUTING_KEY("rabbitmq_ROUTING_KEY", "The routing key that will be used when sending the message"),
```
Any reasons why would Enum would make more sense than pure contents? I don't think description is worth to put it in the Enum which can be a pure Javadoc comment .

rgannu replied:
I changed to Enum to add description. Also I thought from this Enum we can generate the documentation.
If we feel that description is not necessary then I will revert it back.

omarsmak replied:
Since there is no a process I am aware of to generate docs from Enums, I'd revert these to the normal constants.

rgannu replied:
Ok. Reverted to normal constants.

omarsmak commented on components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQMessageConverter.java:
```diff
@@ -49,158 +52,196 @@
             properties.contentType(contentType.toString());
         }
 
-        final Object priority = msg.removeHeader(RabbitMQConstants.PRIORITY);
```
Any reason why this should be mutable by removing `final`?

rgannu replied:
Removed it accidentally. It should be there. Accepted.

oscerd commented:
@davsclaus can you have a look here?
@rgannu can you please resolve conflicts?

omarsmak commented:
Hey @rgannu, the build seems to be failing with this error:
```
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[31,37] error: package org.apache.camel.support.beans does not exist
[ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[211,82] error: cannot find symbol
  symbol:   class SimpleDataHolderBean
  location: class RabbitMQComponent
[ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[213,85] error: cannot find symbol
  symbol:   class SimpleDataHolderBean
  location: class RabbitMQComponent
[INFO] 3 errors
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.436 s
[INFO] Finished at: 2020-10-05T10:46:25+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project camel-rabbitmq: Compilation failure: Compilation failure:
[ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[31,37] error: package org.apache.camel.support.beans does not exist
[ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[211,82] error: cannot find symbol
[ERROR]   symbol:   class SimpleDataHolderBean
[ERROR]   location: class RabbitMQComponent
[ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[213,85] error: cannot find symbol
[ERROR]   symbol:   class SimpleDataHolderBean
[ERROR]   location: class RabbitMQComponent
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```
Can you please take a look at it?

rgannu commented:
> Hey @rgannu, the build seems to be failing with this error:
> 
> ```
> [ERROR] COMPILATION ERROR :
> [INFO] -------------------------------------------------------------
> [ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[31,37] error: package org.apache.camel.support.beans does not exist
> [ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[211,82] error: cannot find symbol
>   symbol:   class SimpleDataHolderBean
>   location: class RabbitMQComponent
> [ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[213,85] error: cannot find symbol
>   symbol:   class SimpleDataHolderBean
>   location: class RabbitMQComponent
> [INFO] 3 errors
> [INFO] -------------------------------------------------------------
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  4.436 s
> [INFO] Finished at: 2020-10-05T10:46:25+02:00
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project camel-rabbitmq: Compilation failure: Compilation failure:
> [ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[31,37] error: package org.apache.camel.support.beans does not exist
> [ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[211,82] error: cannot find symbol
> [ERROR]   symbol:   class SimpleDataHolderBean
> [ERROR]   location: class RabbitMQComponent
> [ERROR] /Users/omaral-safi/Code/camel/components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQComponent.java:[213,85] error: cannot find symbol
> [ERROR]   symbol:   class SimpleDataHolderBean
> [ERROR]   location: class RabbitMQComponent
> [ERROR] -> [Help 1]
> [ERROR]
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR]
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
> ```
> 
> Can you please take a look at it?

Hi @omarsmak 
This is a new class `SimpleDataHolderBean` which I introduced and it is in the separate `package org.apache.camel.support.beans ` .. I have pushed that as well into my branch. So, i am not sure where you are getting this error

rgannu commented:
> @rgannu can you please resolve conflicts?

@oscerd, I resolved this last week.
https://github.com/apache/camel/pull/4346/commits/a6afab81bbe5a81d8b139d894a8e30ad4383e9c7

oscerd commented:
This cannot be rebased, so you'll have to fix them again I believe.

rgannu commented:
> This cannot be rebased, so you'll have to fix them again I believe.

I removed the rebase commits and updated and fixed the conflicts.
Is this ok?

oscerd commented:
It cannot be rebased right now

omarsmak commented on core/camel-support/src/main/java/org/apache/camel/support/beans/SimpleDataHolderBean.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class SimpleDataHolderBean {
```
@rgannu thank you, didn't notice this class in the `org.apache.camel.support.beans` package. However, I am wondering, is this really needed to be in `org.apache.camel.support.beans` package? As I can see here, this can be just moved to the component rabbitmq package, unless there is a reason which I am not aware of?

rgannu replied:
This bean could be used for anything and not just for rabbitmq and hence I moved it inside camel-support module.

davsclaus replied:
I am sorry but we do not want to make camel-core / camel-support bigger than what we think is needed. So can you not do that and add the bean inside the rabbitmq component

rgannu replied:
ok. No problem. I will move it to camel-rabbitmq module this bean.

rgannu commented:
> It cannot be rebased right now

@oscerd, I didn't get you. What I need to do now?
Should I close this PR and open a new one and cherry-pick these

oscerd commented:
First I guess you need to check other reviews. We prefer to avoid that bean.

oscerd commented:
> > It cannot be rebased right now
> 
> @oscerd, I didn't get you. What I need to do now?
> Should I close this PR and open a new one and cherry-pick these

You need to git pull --rebase origin master on your branch and fix conflicts

rgannu commented:
> First I guess you need to check other reviews. We prefer to avoid that bean.

I have moved the bean `SimpleDataHolderBean`  to the `org.apache.camel.component.rabbitmq` packag.

omarsmak commented on core/camel-support/src/main/java/org/apache/camel/component/rabbitmq/SimpleDataHolderBean.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class SimpleDataHolderBean {
```
@rgannu we mentioned that we want this class to be in the component's package which means the component's folder (`components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq`)

rgannu replied:
oops. I wanted to also move into the camel-rabbitmq module only. The refactoring in intellij has moved it to somewhere and I didn't notice. Fixed it properly now. 
Also, as @oscerd was mentioning about the rebase problem I created again a new branch and moved this work to that. 
https://github.com/apache/camel/pull/4371

Is it possible to move the comment from this PR to 4371 PR? I do not know.

rgannu commented:
Created another pull request as there are conflicts problem due to rebase:
https://github.com/apache/camel/pull/4371
